keps/sig-architecture/4402-go-workspaces/README.md

```diff
@@ -0,0 +1,374 @@
+
+The Go project saw what we (and others) were doing and did not like it.  So
+they created [workspaces](https://go.dev/doc/tutorial/workspaces).  They
+basically created a solution that is purpose-built for us.
+
+This KEP proposes that we adopt Go workspaces and bring our tooling up to
+modern standards.  In fact, the author started this work in 2021 or 2022,
+discovering issues along the way that the Go team has worked to resolve.
+
+This is not a user-facing change.  No k8s cluster-user or cluster-admin should
+know or care that this happened.  On the surface, it seems like something that
+should not warrant a KEP, but:
+  a) KEPs are how we communicate big changes;
+  b) This is a big change;
+  c) There's some ecosystem impact to developers
+
+-->
+
+## Motivation
+
+Anyone who has had to maintain the multitude of `verify-*` and `update-*`
+scripts has felt this pain.  Anyone who has had to deal with how we build and
+test kubernetes has run into the mess.  It's a constant source of complexity
+and friction for maintainers, and it leaks out into our developer ecosystem.
+
+This is the sort of mess that wise people back away from.  The author is not so
+wise.
+
+### Goals
+
+1) To get rid of our reliance on GOPATH hackery to build Kubernetes and to run
+   verifier tools and code-generators.
+2) To simplify and clarify our tooling with regards to multi-module operation.
+
+### Non-Goals
+
+1) To rewrite our code-generators (though some could use it).
+2) To make the build or tools faster.
+
+## Proposal
+
+Thie KEP proposes to add a `go.work` file (checked in) and fix all the tools
+that break because of that.  At the end, there will be no reliance on GOPATH.
+verify and update scripts can use normal Go conventions for building and
+running code.  Abominations like `run-in-gopath.sh` will be deleted.
+
+### Risks and Mitigations
+
+There's not a clean way to do this.  Go workspaces DOES NOT work with our
+symlinks-in-vendor hack, and all of our tools do not work without it.  We have
+a LOT of tools and the hacks run DEEP.  We have a decade of legacy to scrub
+off.  Worse, many of our tools make dubious assumptions about the
+interchangeability of Go package paths (e.g. "example.com/foo/bar") and dis
+paths.  Those assumptions held as long as we used GOPATH, BARELY held with
+staging modules, and break totally in workspaces.
+
+This means that some of our ecosystem-facing APIs (codegen tools, in
+particular) *must* change.  We can mitigate some of the impact by creating "v2"
+of things like k8s.io/gengo (which powers many tools).  We might choose to do
+the same in kube-openapi, though its API is not changing, just the
+implementation.
+
+However, k8s.io/code-generator is more challenging.  It holds all of our tools
+and some scripts for invoking them, and is used by a non-trivial number of
+downstream projects (operators, custom API servers, etc).  Making a v2 of this
```

Comment:
and keep it as `legacy-code-generator`? we have already `legacy-cloud-providers` in staging

Reply:
We actively maintain legacy-cloud-providers (minimal but not zero).  This would be zero.  We could do this or we could just tag the last release of k8s.io/code-generator.